### PR TITLE
kubernetes-discovery proxy

### PR DESCRIPTION
**Pull request by deads2k**
The last commit adds an authenticating (but not authorizing) proxy to `kubernetes-discovery`.  The other commits are in the submit queue already.

@sttts @cjcullen This is it.  After this we can start e2e wiring.

### CODE REVIEW

File: cmd/kubernetes-discovery/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -47,6 +46,11 @@ const legacyAPIServiceName = "v1."
 type Config struct {
 	GenericConfig *genericapiserver.Config
 
+	// ProxyClientCert/Key are the client cert use to identify this proxy. Backing APIServices use
```
Comment (sttts) [inline cmd/kubernetes-discovery/pkg/apiserver/apiserver.go]: used

Comment (deads2k) [inline cmd/kubernetes-discovery/pkg/apiserver/apiserver.go]: > used

rebased and done

**Comment by deads2k** [comment]
@k8s-bot gce etcd3 e2e test this

did not start

File: cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	location.Scheme = "https"
+	location.Host = r.getDestinationHost()
+	location.Path = req.URL.Path
+	values := location.Query()
+	for k, vs := range req.URL.Query() {
+		for _, v := range vs {
+			values.Add(k, v)
+		}
+	}
+	location.RawQuery = values.Encode()
```
Comment (smarterclayton) [inline cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go]: Why isn't this just `location.RawQuery = req.URL.Query().Encode()`?

File: cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	if err != nil {
+		return nil, true, err
+	}
+
+	return wrappedRT, true, nil
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
```
Comment (smarterclayton) [inline cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go]: If this is supposed to be merging, call it `merge`

File: cmd/kubernetes-discovery/pkg/apiserver/apiserver.go [hunk 4]
```diff
@@ -102,9 +118,13 @@ func (c completedConfig) New() (*APIDiscoveryServer, error) {
...
 	s := &APIDiscoveryServer{
-		GenericAPIServer:   genericServer,
-		handledAPIServices: sets.String{},
-		lister:             informerFactory.Apiregistration().InternalVersion().APIServices().Lister(),
+		GenericAPIServer: genericServer,
+		contextMapper:    c.GenericConfig.RequestContextMapper,
+		proxyClientCert:  c.ProxyClientCert,
+		proxyClientKey:   c.ProxyClientKey,
+		proxyHandlers:    map[string]*proxyHandler{},
+		lister:           informerFactory.Apiregistration().InternalVersion().APIServices().Lister(),
```
Comment (smarterclayton) [inline cmd/kubernetes-discovery/pkg/apiserver/apiserver.go]: Let me guess - we don't generate camel case names for informers?  `Apiregistration` is like we're not even trying...

File: cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	}
+}
+
+// responder implements rest.Responder for assisting a connector in writing objects or errors.
+type responder struct {
+	w http.ResponseWriter
+}
+
+func (r *responder) Object(statusCode int, obj runtime.Object) {
+	apiserver.WriteRawJSON(statusCode, obj, r.w)
```
Comment (smarterclayton) [inline cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go]: Not sure this is going to work from a generic client - is this even invoked in your code path?

Comment (smarterclayton) [inline cmd/kubernetes-discovery/pkg/apiserver/handler_proxy.go]: You'll need to deal with content negotiation in the aggregator - if the caller asked for protobuf and you write JSON bad things happen.

**Comment by smarterclayton** [comment]
Not dealing with content negotiation is going to be a problem.  At least add todo's around all the places you make specific assumptions about errors, response types, and negotiation headers.

**Comment by deads2k** [comment]
@smarterclayton made it through and ok otherwise?

**Pull request merged**